### PR TITLE
Add progress bar control

### PR DESCRIPTION
**Pull request by coolreader18**
libui has had a ProgressBar control for a while, and this PR would implement it.

### Stability

As far as I know, ProgressBar is stable. If it isn't, I don't know where to look for stability information for libui.

### CODE REVIEW

File: iui/src/controls/progressbar.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    pub fn set_indeterminate(&mut self) {
+        self.set_value(ProgressBarValue::Indeterminate);
+    }
+
+    /// Set the value of the progress bar
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the value is `Determinate` and its value is not
+    /// in the range 0..=100
```
Comment (NoraCodes) [inline iui/src/controls/progressbar.rs]: I would prefer to simply cap this value; if a user provides a value of <0 it should be set to 0, and a value of >100 should be set to 100.

**Review by NoraCodes — CHANGES_REQUESTED** [review]
Please remove the formatting changes in the changelog and address one comment; aside from that, looks good. Thanks a lot!

File: iui/src/controls/progressbar.rs (hunk 1 above)
Comment (coolreader18) [reply iui/src/controls/progressbar.rs]: Sounds good!

File: iui/src/controls/progressbar.rs [hunk 2]
```diff
@@ -7,7 +7,7 @@ use ui_sys::{self, uiControl, uiProgressBar};
 pub enum ProgressBarValue {
     /// Represents a set, consistent percentage of the bar to be filled
     ///
-    /// The u32 should be in the range 0..=100
+    /// The value should be in the range 0..=100
```
Comment (NoraCodes) [inline iui/src/controls/progressbar.rs]: Might be good to mention here what will happen if the value is not in that range.

File: iui/src/controls/progressbar.rs [hunk 3]
```diff
@@ -40,10 +40,6 @@ impl ProgressBar {
     }
 
     /// Set the value of the progress bar to a determinate value
-    ///
```
Comment (NoraCodes) [inline iui/src/controls/progressbar.rs]: Might want to add a note here, something like: "Values out of the range 0..=100 will be capped to that range"

File: iui/src/controls/progressbar.rs [hunk 4]
```diff
@@ -54,18 +50,10 @@ impl ProgressBar {
...
-    /// This function will panic if the value is `Determinate` and its value is not
-    /// in the range 0..=100
     pub fn set_value(&mut self, value: ProgressBarValue) {
         let sys_value = match value {
             ProgressBarValue::Determinate(value) => {
-                assert!(match value {
-                    0..=100 => true,
-                    _ => false,
-                });
+                let value = if value > 100 { 100 } else { value };
```
Comment (NoraCodes) [inline iui/src/controls/progressbar.rs:59]: Values less than zero can still be entered, which will cause either UB or a panic with your assert about negative values later. Perhaps add a clause to cap negatives to 0?

**Review by NoraCodes — CHANGES_REQUESTED** [review]

File: iui/src/controls/progressbar.rs (hunk 4 above)
Comment (coolreader18) [reply iui/src/controls/progressbar.rs:59]: Values less than 0 can't be passed because `Determinate` holds a u32.

Comment (NoraCodes) [reply iui/src/controls/progressbar.rs:59]: Oh, duh. My bad. Good to go then, pending CI success.

**Comment by NoraCodes** [comment]
Thanks!

**Pull request merged**